app/services/foreman/client_certificate.rb

```diff
@@ -0,0 +1,56 @@
+module Foreman
+  class ClientCertificate
+    delegate :logger, to: :Rails
+    attr_reader :request
+
+    def initialize(request:)
+      @request = request
+    end
+
+    def present?
+      @present ||= request.ssl? && request.env.has_key?(certificate_key) && raw_data.present? && raw_data != '(null)'
+    end
+
+    def raw_data
+      request.env[certificate_key]
+    end
+
+    def subject
+      return certificate_extract.subject if present?
+      dn = request.env[dn_key]
+      return unless dn && dn =~ /CN=([^\s\/,]+)/i
+
+      Regexp.last_match(1).downcase
+    end
+
+    def subject_alternative_names
+      return unless present?
+      certificate_extract.subject_alternative_names
+    end
+
+    def verified?
+      verify = request.env[verify_key]
+      return true if verify == 'SUCCESS'
+      logger.info { "Client certificate is invalid: #{verify}" }
+      false
+    end
+
+    private
+
+    def certificate_key
```

Comment:
key is a bit ambiguous here since it can be read as the private key.
```suggestion
    def certificate_env_key
```